File: mcp/jira-mcp/src/services/processMyTimelogs.ts
import type { JiraAuthType, JiraIssueSearchItem, JiraUser, JiraWorklog } from '../types/index.js';
import { getCurrentJiraUser, getJiraIssueWorklogs, searchJiraIssues } from './jiraApiClient.js';

const DAY_IN_MS = 24 * 60 * 60 * 1000;

const formatDuration = (totalSeconds: number): string => {
  if (totalSeconds <= 0) {
    return '0m';
  }

  const hours = Math.floor(totalSeconds / 3600);
  const minutes = Math.floor((totalSeconds % 3600) / 60);

  if (hours === 0) {
    return `${minutes}m`;
  }

  if (minutes === 0) {
    return `${hours}h`;
  }

  return `${hours}h ${minutes}m`;
};

const extractTextFromAdfNode = (value: unknown): string => {
  if (typeof value === 'string') {
    return value;
  }

  if (!value || typeof value !== 'object') {
    return '';
  }

  const node = value as {
    text?: unknown;
    content?: unknown;
  };

  const text = typeof node.text === 'string' ? node.text : '';
  const content = Array.isArray(node.content)
    ? node.content
        .map((item) => extractTextFromAdfNode(item))
        .filter(Boolean)
        .join(' ')
    : '';

  return `${text} ${content}`.trim();
};

const getCommentText = (comment: unknown): string => {
  if (!comment) {
    return '';
  }

  if (typeof comment === 'string') {
    return comment.trim();
  }

  return extractTextFromAdfNode(comment).replace(/\s+/g, ' ').trim();
};

const isSameUser = (user: JiraUser, author: JiraUser | undefined): boolean => {
  if (!author) {
    return false;
  }

  if (user.accountId && author.accountId) {
    return user.accountId === author.accountId;
  }

  if (user.emailAddress && author.emailAddress) {
    return user.emailAddress.toLowerCase() === author.emailAddress.toLowerCase();
  }

  if (user.displayName && author.displayName) {
    return user.displayName.toLowerCase() === author.displayName.toLowerCase();
  }

  return false;
};

type TimelogEntry = {
  started: string;
  date: string;
  projectKey: string;
  projectName: string;
  issueKey: string;
  issueSummary: string;
  timeSpentSeconds: number;
  timeSpent: string;
  workDescription: string;
};

type TimelogIssueSummary = {
  issueKey: string;
  issueSummary: string;
  totalTimeSeconds: number;
  totalTime: string;
  entries: TimelogEntry[];
};

type TimelogProjectSummary = {
  projectKey: string;
  projectName: string;
  totalTimeSeconds: number;
  totalTime: string;
  issues: TimelogIssueSummary[];
};

export type TimelogReport = {
  period: {
    days: number;
    from: string;
    to: string;
  };
  author: {
    accountId: string;
    displayName: string;
    emailAddress: string;
  };
  totalTimeSeconds: number;
  totalTime: string;
  entriesCount: number;
  projects: TimelogProjectSummary[];
  details: TimelogEntry[];
  markdownSummary: string;
};

const buildMarkdownSummary = (report: Omit<TimelogReport, 'markdownSummary'>): string => {
  const lines: string[] = [];

  lines.push('# Jira timelog summary');
  lines.push('');
  lines.push(`- Period: ${report.period.from} to ${report.period.to} (${report.period.days} days)`);
  lines.push(`- User: ${report.author.displayName || 'Unknown user'}`);
  lines.push(`- Total logged: ${report.totalTime} across ${report.projects.length} project(s) and ${report.entriesCount} worklog entr${report.entriesCount === 1 ? 'y' : 'ies'}`);
  lines.push('');

  if (report.projects.length === 0) {
    lines.push('No timelogs found for the selected period.');
    return lines.join('\n');
  }

  lines.push('## Projects');
  lines.push('');

  for (const project of report.projects) {
    lines.push(`### ${project.projectKey} - ${project.projectName} (${project.totalTime})`);
    for (const issue of project.issues) {
      lines.push(`- ${issue.issueKey} (${issue.totalTime}): ${issue.issueSummary}`);
    }
    lines.push('');
  }

  lines.push('## Detailed work');
  lines.push('');

  for (const item of report.details) {
    lines.push(`- ${item.date} | ${item.projectKey} | ${item.issueKey} | ${item.timeSpent} | ${item.workDescription}`);
  }

  return lines.join('\n').trim();
};

const getAllCandidateIssues = async ({
  jiraBaseUrl,
  jiraApiToken,
  jiraEmail,
  jiraAuthType,
  days,
}: {
  jiraBaseUrl: string;
  jiraApiToken: string;
  jiraEmail: string;
  jiraAuthType: JiraAuthType;
  days: number;
}): Promise<JiraIssueSearchItem[]> => {
  const jql = `worklogAuthor = currentUser() AND worklogDate >= -${days}d ORDER BY updated DESC`;
  const issues: JiraIssueSearchItem[] = [];
  let startAt = 0;

  while (true) {
    const page = await searchJiraIssues({
      baseUrl: jiraBaseUrl,
      token: jiraApiToken,
      email: jiraEmail,
      jiraAuthType,
      jql,
      fields: ['summary', 'project'],
      maxResults: 100,
      startAt,
    });

    const pageIssues = page.issues || [];
    issues.push(...pageIssues);

    const pageSize = page.maxResults && page.maxResults > 0 ? page.maxResults : pageIssues.length;
    const nextStartAt = startAt + pageSize;
    const total = page.total || issues.length;

    if (pageIssues.length === 0 || nextStartAt >= total || pageSize === 0) {
      break;
    }

    startAt = nextStartAt;
  }

  return issues;
};

const getIssueWorklogs = async ({
  jiraBaseUrl,
  jiraApiToken,
  jiraEmail,
  jiraAuthType,
  issueKey,
}: {
  jiraBaseUrl: string;
  jiraApiToken: string;
  jiraEmail: string;
  jiraAuthType: JiraAuthType;
  issueKey: string;
}): Promise<JiraWorklog[]> => {
  const worklogs: JiraWorklog[] = [];
  let startAt = 0;

  while (true) {
    const page = await getJiraIssueWorklogs({
      baseUrl: jiraBaseUrl,
      token: jiraApiToken,
      email: jiraEmail,
      jiraAuthType,
      issueKey,
      maxResults: 100,
      startAt,
    });

    const pageWorklogs = page.worklogs || [];
    worklogs.push(...pageWorklogs);

    const pageSize = page.maxResults && page.maxResults > 0 ? page.maxResults : pageWorklogs.length;
    const nextStartAt = startAt + pageSize;
    const total = page.total || worklogs.length;

    if (pageWorklogs.length === 0 || nextStartAt >= total || pageSize === 0) {
      break;
    }

    startAt = nextStartAt;
  }

  return worklogs;
};

type ProcessMyTimelogsInput = {
  jiraBaseUrl: string;
  jiraApiToken: string;
  jiraEmail: string;
  jiraAuthType: JiraAuthType;
  days: number;
};

export const processMyTimelogs = async ({
  jiraBaseUrl,
  jiraApiToken,
  jiraEmail,
  jiraAuthType,
  days,
}: ProcessMyTimelogsInput): Promise<TimelogReport> => {
  const now = new Date();
  const periodStart = new Date(now.getTime() - days * DAY_IN_MS);
  const me = await getCurrentJiraUser({
    baseUrl: jiraBaseUrl,
    token: jiraApiToken,
    email: jiraEmail,
    jiraAuthType,
  });

  const candidateIssues = await getAllCandidateIssues({
    jiraBaseUrl,
    jiraApiToken,
    jiraEmail,
    jiraAuthType,
    days,
  });

  const details: TimelogEntry[] = [];

  for (const issue of candidateIssues) {
    if (!issue.key) {
      continue;
    }

    const issueKey = issue.key;
    const issueSummary = issue.fields?.summary || 'No issue summary';
    const projectKey = issue.fields?.project?.key || 'UNKNOWN';
    const projectName = issue.fields?.project?.name || 'Unknown project';

    const issueWorklogs = await getIssueWorklogs({
      jiraBaseUrl,
      jiraApiToken,
      jiraEmail,
      jiraAuthType,
      issueKey,
    });

    for (const worklog of issueWorklogs) {
      const started = worklog.started;
      const timeSpentSeconds = worklog.timeSpentSeconds || 0;

      if (!started || timeSpentSeconds <= 0) {
        continue;
      }

      const startedAt = new Date(started);
      if (Number.isNaN(startedAt.getTime()) || startedAt < periodStart || startedAt > now) {
        continue;
      }

      if (!isSameUser(me, worklog.author)) {
        continue;
      }

      const commentText = getCommentText(worklog.comment);

      details.push({
        started,
        date: started.slice(0, 10),
        projectKey,
        projectName,
        issueKey,
        issueSummary,
        timeSpentSeconds,
        timeSpent: formatDuration(timeSpentSeconds),
        workDescription: commentText || issueSummary,
      });
    }
  }

  details.sort((a, b) => new Date(b.started).getTime() - new Date(a.started).getTime());

  const projectsMap = new Map<string, TimelogProjectSummary>();
  let totalTimeSeconds = 0;

  for (const detail of details) {
    totalTimeSeconds += detail.timeSpentSeconds;

    const projectId = `${detail.projectKey}::${detail.projectName}`;
    const existingProject = projectsMap.get(projectId);
    const project =
      existingProject ||
      {
        projectKey: detail.projectKey,
        projectName: detail.projectName,
        totalTimeSeconds: 0,
        totalTime: '0m',
        issues: [],
      };

    project.totalTimeSeconds += detail.timeSpentSeconds;

    let issue = project.issues.find((item) => item.issueKey === detail.issueKey);
    if (!issue) {
      issue = {
        issueKey: detail.issueKey,
        issueSummary: detail.issueSummary,
        totalTimeSeconds: 0,
        totalTime: '0m',
        entries: [],
      };
      project.issues.push(issue);
    }

    issue.totalTimeSeconds += detail.timeSpentSeconds;
    issue.entries.push(detail);
    projectsMap.set(projectId, project);
  }

  const projects = [...projectsMap.values()]
    .map((project) => ({
      ...project,
      totalTime: formatDuration(project.totalTimeSeconds),
      issues: project.issues
        .map((issue) => ({
          ...issue,
          totalTime: formatDuration(issue.totalTimeSeconds),
          entries: issue.entries.sort((a, b) => new Date(b.started).getTime() - new Date(a.started).getTime()),
        }))
        .sort((a, b) => b.totalTimeSeconds - a.totalTimeSeconds),
    }))
    .sort((a, b) => b.totalTimeSeconds - a.totalTimeSeconds);

  const reportWithoutMarkdown: Omit<TimelogReport, 'markdownSummary'> = {
    period: {
      days,
      from: periodStart.toISOString().slice(0, 10),
      to: now.toISOString().slice(0, 10),
    },
    author: {
      accountId: me.accountId || '',
      displayName: me.displayName || 'Unknown user',
      emailAddress: me.emailAddress || '',
    },
    totalTimeSeconds,
    totalTime: formatDuration(totalTimeSeconds),
    entriesCount: details.length,
    projects,
    details,
  };

  return {
    ...reportWithoutMarkdown,
    markdownSummary: buildMarkdownSummary(reportWithoutMarkdown),
  };
};

File: mcp/jira-mcp/src/jira_mcp/services/process_my_timelogs.py
from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import Any

from jira_mcp.config import JiraAuthType
from jira_mcp.services.jira_api_client import (
    get_current_jira_user,
    get_jira_issue_worklogs,
    search_jira_issues,
)

JsonDict = dict[str, Any]


def _format_duration(total_seconds: int) -> str:
    if total_seconds <= 0:
        return "0m"

    hours = total_seconds // 3600
    minutes = (total_seconds % 3600) // 60

    if hours == 0:
        return f"{minutes}m"
    if minutes == 0:
        return f"{hours}h"
    return f"{hours}h {minutes}m"


def _extract_text_from_adf_node(value: Any) -> str:
    if isinstance(value, str):
        return value
    if not isinstance(value, dict):
        return ""

    text = value.get("text") if isinstance(value.get("text"), str) else ""
    content = value.get("content")
    if not isinstance(content, list):
        return text.strip()

    nested = " ".join(_extract_text_from_adf_node(item) for item in content).strip()
    return f"{text} {nested}".strip()


def _get_comment_text(comment: Any) -> str:
    if not comment:
        return ""
    if isinstance(comment, str):
        return comment.strip()
    return " ".join(_extract_text_from_adf_node(comment).split())


def _is_same_user(user: JsonDict, author: JsonDict | None) -> bool:
    if not author:
        return False

    user_account_id = user.get("accountId")
    author_account_id = author.get("accountId")
    if user_account_id and author_account_id:
        return user_account_id == author_account_id

    user_email = user.get("emailAddress")
    author_email = author.get("emailAddress")
    if isinstance(user_email, str) and isinstance(author_email, str):
        return user_email.lower() == author_email.lower()

    user_name = user.get("displayName")
    author_name = author.get("displayName")
    if isinstance(user_name, str) and isinstance(author_name, str):
        return user_name.lower() == author_name.lower()

    return False


def _parse_jira_datetime(value: str) -> datetime | None:
    try:
        normalized = value
        if len(normalized) >= 5 and normalized[-5] in {"+", "-"} and normalized[-3] != ":":
            normalized = f"{normalized[:-2]}:{normalized[-2:]}"
        return datetime.fromisoformat(normalized)
    except ValueError:
        return None


def _sort_timestamp(value: str) -> float:
    parsed = _parse_jira_datetime(value)
    if not parsed:
        return 0.0
    return parsed.timestamp()


def _build_markdown_summary(report: JsonDict) -> str:
    lines: list[str] = []
    period = report["period"]
    author = report["author"]
    projects = report["projects"]
    details = report["details"]
    entries_count = report["entriesCount"]

    lines.append("# Jira timelog summary")
    lines.append("")
    lines.append(f"- Period: {period['from']} to {period['to']} ({period['days']} days)")
    lines.append(f"- User: {author['displayName'] or 'Unknown user'}")
    lines.append(
        f"- Total logged: {report['totalTime']} across {len(projects)} project(s) "
        f"and {entries_count} worklog entr{'y' if entries_count == 1 else 'ies'}"
    )
    lines.append("")

    if not projects:
        lines.append("No timelogs found for the selected period.")
        return "\n".join(lines)

    lines.append("## Projects")
    lines.append("")
    for project in projects:
        lines.append(f"### {project['projectKey']} - {project['projectName']} ({project['totalTime']})")
        for issue in project["issues"]:
            lines.append(f"- {issue['issueKey']} ({issue['totalTime']}): {issue['issueSummary']}")
        lines.append("")

    lines.append("## Detailed work")
    lines.append("")
    for item in details:
        lines.append(
            f"- {item['date']} | {item['projectKey']} | {item['issueKey']} | "
            f"{item['timeSpent']} | {item['workDescription']}"
        )

    return "\n".join(lines).strip()


async def _get_all_candidate_issues(
    *,
    jira_base_url: str,
    jira_api_token: str,
    jira_email: str,
    jira_auth_type: JiraAuthType,
    days: int,
) -> list[JsonDict]:
    jql = f"worklogAuthor = currentUser() AND worklogDate >= -{days}d ORDER BY updated DESC"
    issues: list[JsonDict] = []
    start_at = 0

    while True:
        page = await search_jira_issues(
            base_url=jira_base_url,
            token=jira_api_token,
            email=jira_email,
            jira_auth_type=jira_auth_type,
            jql=jql,
            fields=["summary", "project"],
            start_at=start_at,
            max_results=100,
        )

        page_issues = page.get("issues") or []
        issues.extend(page_issues)

        page_size = int(page.get("maxResults") or len(page_issues))
        next_start_at = start_at + page_size
        total = int(page.get("total") or len(issues))

        if not page_issues or next_start_at >= total or page_size == 0:
            break
        start_at = next_start_at

    return issues


async def _get_issue_worklogs(
    *,
    jira_base_url: str,
    jira_api_token: str,
    jira_email: str,
    jira_auth_type: JiraAuthType,
    issue_key: str,
) -> list[JsonDict]:
    worklogs: list[JsonDict] = []
    start_at = 0

    while True:
        page = await get_jira_issue_worklogs(
            base_url=jira_base_url,
            token=jira_api_token,
            email=jira_email,
            jira_auth_type=jira_auth_type,
            issue_key=issue_key,
            start_at=start_at,
            max_results=100,
        )

        page_worklogs = page.get("worklogs") or []
        worklogs.extend(page_worklogs)

        page_size = int(page.get("maxResults") or len(page_worklogs))
        next_start_at = start_at + page_size
        total = int(page.get("total") or len(worklogs))

        if not page_worklogs or next_start_at >= total or page_size == 0:
            break
        start_at = next_start_at

    return worklogs


async def process_my_timelogs(
    *,
    jira_base_url: str,
    jira_api_token: str,
    jira_email: str,
    jira_auth_type: JiraAuthType,
    days: int,
) -> JsonDict:
    now = datetime.now(timezone.utc)
    period_start = now - timedelta(days=days)

    me = await get_current_jira_user(
        base_url=jira_base_url,
        token=jira_api_token,
        email=jira_email,
        jira_auth_type=jira_auth_type,
    )

    candidate_issues = await _get_all_candidate_issues(
        jira_base_url=jira_base_url,
        jira_api_token=jira_api_token,
        jira_email=jira_email,
        jira_auth_type=jira_auth_type,
        days=days,
    )

    details: list[JsonDict] = []
    for issue in candidate_issues:
        issue_key = issue.get("key")
        if not issue_key:
            continue

        issue_fields = issue.get("fields") or {}
        issue_summary = issue_fields.get("summary") or "No issue summary"
        project = issue_fields.get("project") or {}
        project_key = project.get("key") or "UNKNOWN"
        project_name = project.get("name") or "Unknown project"

        issue_worklogs = await _get_issue_worklogs(
            jira_base_url=jira_base_url,
            jira_api_token=jira_api_token,
            jira_email=jira_email,
            jira_auth_type=jira_auth_type,
            issue_key=issue_key,
        )

        for worklog in issue_worklogs:
            started = worklog.get("started")
            time_spent_seconds = int(worklog.get("timeSpentSeconds") or 0)
            if not isinstance(started, str) or time_spent_seconds <= 0:
                continue

            started_at = _parse_jira_datetime(started)
            if not started_at or started_at < period_start or started_at > now:
                continue

            if not _is_same_user(me, worklog.get("author")):
                continue

            comment_text = _get_comment_text(worklog.get("comment"))
            details.append(
                {
                    "started": started,
                    "date": started[:10],
                    "projectKey": project_key,
                    "projectName": project_name,
                    "issueKey": issue_key,
                    "issueSummary": issue_summary,
                    "timeSpentSeconds": time_spent_seconds,
                    "timeSpent": _format_duration(time_spent_seconds),
                    "workDescription": comment_text or issue_summary,
                }
            )

    details.sort(key=lambda item: _sort_timestamp(item["started"]), reverse=True)

    projects_map: dict[str, JsonDict] = {}
    total_time_seconds = 0

    for detail in details:
        total_time_seconds += int(detail["timeSpentSeconds"])
        project_id = f"{detail['projectKey']}::{detail['projectName']}"

        project = projects_map.get(project_id)
        if not project:
            project = {
                "projectKey": detail["projectKey"],
                "projectName": detail["projectName"],
                "totalTimeSeconds": 0,
                "totalTime": "0m",
                "issues": [],
            }

        project["totalTimeSeconds"] = int(project["totalTimeSeconds"]) + int(detail["timeSpentSeconds"])

        issue_summary = None
        for item in project["issues"]:
            if item["issueKey"] == detail["issueKey"]:
                issue_summary = item
                break

        if not issue_summary:
            issue_summary = {
                "issueKey": detail["issueKey"],
                "issueSummary": detail["issueSummary"],
                "totalTimeSeconds": 0,
                "totalTime": "0m",
                "entries": [],
            }
            project["issues"].append(issue_summary)

        issue_summary["totalTimeSeconds"] = int(issue_summary["totalTimeSeconds"]) + int(detail["timeSpentSeconds"])
        issue_summary["entries"].append(detail)
        projects_map[project_id] = project

    projects: list[JsonDict] = []
    for project in projects_map.values():
        project["totalTime"] = _format_duration(int(project["totalTimeSeconds"]))
        normalized_issues: list[JsonDict] = []
        for issue in project["issues"]:
            issue["totalTime"] = _format_duration(int(issue["totalTimeSeconds"]))
            issue["entries"].sort(
                key=lambda item: _sort_timestamp(item["started"]),
                reverse=True,
            )
            normalized_issues.append(issue)
        normalized_issues.sort(key=lambda item: int(item["totalTimeSeconds"]), reverse=True)
        project["issues"] = normalized_issues
        projects.append(project)

    projects.sort(key=lambda item: int(item["totalTimeSeconds"]), reverse=True)

    report: JsonDict = {
        "period": {
            "days": days,
            "from": period_start.date().isoformat(),
            "to": now.date().isoformat(),
        },
        "author": {
            "accountId": me.get("accountId") or "",
            "displayName": me.get("displayName") or "Unknown user",
            "emailAddress": me.get("emailAddress") or "",
        },
        "totalTimeSeconds": total_time_seconds,
        "totalTime": _format_duration(total_time_seconds),
        "entriesCount": len(details),
        "projects": projects,
        "details": details,
    }
    report["markdownSummary"] = _build_markdown_summary(report)
    return report
